### PR TITLE
fix(functorch): Fix vjp_fn behavior when compiled with torch.compile

### LINKED ISSUE #172386: torch.func.vjp: compiled vjp_fn returns incorrect zero gradients
### Description

When compiling the function returned by `torch.func.vjp` using torch.compile, the compiled vjp_fn returns incorrect gradients (all zeros).

This is a silent correctness issue:
* eager vjp_fn returns correct gradients
* compiled vjp_fn runs without error
* compiled output is wrong

### Minimal Reproducible Example

```python
import torch
from torch import nn
from torch.func import vjp, functional_call

torch.manual_seed(0)

model = nn.Sequential(
    nn.Linear(4, 1)
)

x = torch.randn(4, 4)
params = {k: v for k, v in model.named_parameters()}

# Functional form
f = lambda p: functional_call(model, p, x)

# Eager VJP works
y, vjp_fn = vjp(f, params)
v = torch.ones_like(y)
gx = vjp_fn(v)[0]
print("eager vjp:", gx) # Non-zero

# Compile the vjp function
vjp_fn_compiled = torch.compile(vjp_fn)
gx2 = vjp_fn_compiled(v)[0]
print("compiled vjp:", gx2)  # all zeros
```

### Observed behavior

* vjp_fn(v) (eager) returns correct non-zero gradients
* torch.compile(vjp_fn)(v) returns a dictionary of all zeros
* No error or warning is raised

### Expected behavior

The compiled vjp_fn should:
* return the same gradients as the eager version
* raise an error

### Error logs

No error is raised

### Versions

PyTorch version: 2.9.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.6.3.2)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 12:22:05) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-26.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.3.3
[pip3] torch==2.9.1
[pip3] torch-dct==0.1.6
[pip3] torchaudio==2.9.1
[pip3] torchmetrics==1.8.2
[pip3] torchvision==0.24.1
[conda] numpy                        2.3.3            pypi_0           pypi
[conda] torch                        2.9.1            pypi_0           pypi
[conda] torch-dct                    0.1.6            pypi_0           pypi
[conda] torchaudio                   2.9.1            pypi_0           pypi
[conda] torchmetrics                 1.8.2            pypi_0           pypi
[conda] torchvision                  0.24.1           pypi_0           pypi

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @Chillee @samdow @kshitij12345

### PR DESCRIPTION
**Pull request by ayushsatyam146**
**Summary** 
Fixes #172386. `torch.compile()` on vjp functions returned by `torch.func.vjp()` is now consistent with eager execution

**Problem**
When calling `torch.func.vjp(f, x)` in eager mode, we get back `(primals_out, vjp_fn).` The vjp_fn internally captures flat_primals_out (the forward outputs) which have a grad_fn attached for autograd to compute gradients.
However, when you try to compile this `vjp_fn`:

```
_, vjp_fn = torch.func.vjp(f, x)
compiled_vjp = torch.compile(vjp_fn)
compiled_vjp(v)  # Returns zeros or incorrect gradients!
```
AOTAutograd traces the function using FakeTensors, and the original `grad_fn` from eager execution becomes invalid in the traced region. The `_autograd_grad` call then operates on tensors without a proper computation graph, returning incorrect results.

**Solution**
We need a fresh computation graph when compiling, not the stale one from eager. So we capture func and primals upfront, and when `is_compiling`() is true, we simply re-run the forward pass inside the traced region. This gives us outputs with a valid `grad_fn` that autograd can actually use.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98

### CODE REVIEW

**Comment by parsshar-RH** [comment]
@ayushsatyam146 
Please check the failing CI checks

File: test/dynamo/test_higher_order_ops.py [hunk 1]
```diff
@@ -6154,6 +6154,41 @@ def get_vjp(v):
         ):
             torch.func.vjp(get_vjp, x)
 
+    def test_compile_vjp_fn_returns_correct_gradients(self):
```
Comment (Skylion007) [inline test/dynamo/test_higher_order_ops.py]: Do we need to fix jvp_fn too?

File: torch/_functorch/eager_transforms.py [hunk 2]
```diff
@@ -434,13 +443,43 @@ def wrapper(
...
                 )
-            result = _autograd_grad(
-                flat_primals_out,
-                flat_diff_primals,
-                flat_cotangents,
-                retain_graph=retain_graph,
-                create_graph=create_graph,
-            )
+
+            if is_compiling():
```
Comment (zou3519) [inline torch/_functorch/eager_transforms.py]: why does this fix the issue?

**Comment by zou3519** [comment]
in general i prefer a solution that doesn't switch on is_compiling()

File: torch/_functorch/eager_transforms.py (hunk 2 above)
Comment (ayushsatyam146) [reply torch/_functorch/eager_transforms.py]: Hi @zou3519 The main problem here is that when vjp_fn is compiled separately from vjp(), the captured flat_primals_out tensors lose their grad_fn during AOT tracing because they become FakeTensors without autograd history. When _autograd_grad() is called on these tensors, it finds no gradient path and returns zeros.

Here it detect this case when compilation is taking place, we re-run the forward pass `captured_func(*captured_primals)` inside the compiled context. This creates fresh outputs with valid grad_fn that `_autograd_grad()` can traverse. 
I also added `vjp_called_inside_compile` flag: so that when vjp() itself is called during compilation, the captured tensors are already valid symbolic tensors and hence no recompute is needed.

**Comment by ayushsatyam146** [comment]
Hi everyone, sorry for a very long pause on this one but after @ezyang's comment I tried to rewrok the approach towards a more Dynamo native solution. 
I have dropped the `is_compiling` approach where I had to do custom handling for a individual function like vjp. 
Now, I have added a Dynamo handler for functorch_autograd_grad similar to [handle_autograd_grad](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/torch.py#L2125) that detects when closure-captured tensors have lost their grad_fn during tracing and graph-breaks instead of silently producing zeros. 
I can look towards a solution where even the graph break is not needed at all, but that would be a much larger change in how dynamo handles closure tracing. 

cc: @ezyang @zou3519 @Lucaskabela @Skylion007 @williamwen42

File: torch/_dynamo/variables/torch.py [hunk 3]
```diff
@@ -375,6 +375,31 @@ def _collect_tensors_with_sources(
     return results
 
 
+def _collect_placeholder_nodes(var: "VariableTracker") -> list[torch.fx.Node]:
+    """
+    NOTE [Detecting lost autograd linkage in closure-captured tensors]:
+
+    Functorch transforms (vjp, grad, jacrev) return closures capturing tensors
```
Comment (morrison-turnansky) [inline torch/_dynamo/variables/torch.py]: Can you make this doc string more clear? Doc string refers to comparing example vs example_value but no comparison occurs. Maybe make it clear that this collection is then meant to be used in a comparison

File: torch/_dynamo/variables/torch.py [hunk 4]
```diff
@@ -375,6 +375,29 @@ def _collect_tensors_with_sources(
     return results
 
 
+def _collect_placeholder_nodes(var: "VariableTracker") -> list[torch.fx.Node]:
+    """Recursively collect FX placeholder nodes from a VariableTracker.
+
+    Handles TensorVariable and BaseListVariable; other types are ignored.
```
Comment (Lucaskabela) [inline torch/_dynamo/variables/torch.py]: Why are these the only two types we handle? This is not obvious to me

File: torch/_dynamo/variables/torch.py [hunk 5]
```diff
@@ -2376,6 +2399,62 @@ def fn(x):
                 ),
             )
 
+        @register(torch._functorch.eager_transforms._autograd_grad)
+        def handle_functorch_autograd_grad(
+            self, tx: "InstructionTranslator", *args, **kwargs
```
Comment (Lucaskabela) [inline torch/_dynamo/variables/torch.py]: nit: need annotaitons for args/kwargs

File: torch/_dynamo/variables/torch.py [hunk 6]
```diff
@@ -2376,6 +2399,62 @@ def fn(x):
...
+            tensors become graph placeholders whose FakeTensors lose grad_fn,
+            causing _autograd_grad to silently return zeros.
+
+            _collect_placeholder_nodes gathers placeholder nodes from the
+            outputs/inputs args. For each, we compare grapharg.example (the real
+            tensor, retains grad_fn) against example_value (FakeTensor, grad_fn
+            is None). A mismatch means autograd linkage was lost, so we
+            graph-break.
+            """
+            if len(args) >= 2:
```
Comment (Lucaskabela) [inline torch/_dynamo/variables/torch.py]: What do we expect kwargs to be here? None?

**Review by Lucaskabela — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/torch.py (hunk 6 above)
Comment (ayushsatyam146) [reply torch/_dynamo/variables/torch.py]: In case of this particualr bug kwargs are `retain_graph, create_graph, and grad_outputs` (from the _autograd_grad signature). This handler is a pre-check that either detects lost linkage and graph-breaks, or returns None to fall through to the default wrap_fx_proxy path in call_function, none of the kwargs change whether grad_fn linkage was lost. For better clarity I've added a docstring line making this explicit and easy to understand.

File: torch/_dynamo/variables/torch.py (hunk 4 above)
Comment (ayushsatyam146) [reply torch/_dynamo/variables/torch.py]: I put these two only since practically `_autograd_grad`'s arguments always come from tree_flatten() (list) and _as_tuple() (tuple) hence BaseListVariable and TensorVariable cover these actual cases. 
But silently ignoring unknown types is wrong from correctness perspective. I've corrected this now and followed the `_collect_tensors_with_sources pattern` by adding `LazyVariableTracker` handling and a final escape via a graph-break on unexpected types instead of silently passing.

Comment (ayushsatyam146) [reply torch/_dynamo/variables/torch.py]: I have tried to fix it @Lucaskabela. Please take a look, Thanks!

File: torch/_dynamo/variables/torch.py [hunk 7]
```diff
@@ -2377,6 +2414,68 @@ def fn(x):
...
             )
 
+        @register(torch._functorch.eager_transforms._autograd_grad)
+        def handle_functorch_autograd_grad(
+            self,
+            tx: "InstructionTranslator",
+            *args: VariableTracker,
+            **kwargs: VariableTracker,
+        ) -> VariableTracker | None:
+            """Graph-break when closure-captured tensors lost their grad_fn.
```
Comment (Lucaskabela) [inline torch/_dynamo/variables/torch.py]: nit: "lost" is past tense - should be "lose"

**Review by Lucaskabela — CHANGES_REQUESTED** [review]
One minor nit from my part - my only other question is related to the overhead cost of realizing all the placeholder variables/tensor variables.

Do we have idea of how this affects our tracing overhead? Can we run a quick experiment if not?

File: torch/_dynamo/variables/torch.py (hunk 7 above)
Comment (ayushsatyam146) [reply torch/_dynamo/variables/torch.py]: I will correct this

**Review by Lucaskabela — APPROVED** [review]
Thanks for the iteration here :) that cost seems acceptable to me
